text/2017-11-05-deque.md

```diff
@@ -0,0 +1,151 @@
+
+`crossbeam-deque` will use the new Crossbeam's epoch-based garbage collector.
+
+# Detailed design
+
+The basis for implementation is the deque from Coco that already had a good run in
+Rayon and consequently in Stylo. However, there will be a few differences:
+
+1. Memory management is done using `crossbeam-epoch`.
+2. The `Worker`/`Stealer` interface is redesigned as `Deque`/`Stealer`.
+3. Several new convenience methods are introduced.
+
+Other than that, the crucial (and the most tricky to implement) methods like
+`push`, `pop`, and `steal` do not depart from the original implementation
+(except for the changes associated with porting it to Crossbeam's new epoch-based GC).
+
+## The interface
+
+```rust
+/// The "worker" side of the deque.
+pub struct Deque<T>;
+
+unsafe impl<T: Send> Send for Deque<T> {}
+
+impl<T> Deque<T> {
+    /// Creates a new deque.
+    pub fn new() -> Deque<T>;
+
+    /// Creates a new deque with specified minimum capacity.
+    pub fn with_min_capacity(min_cap: usize) -> Deque<T>;
+
+    /// Returns `true` if the deque is empty.
+    pub fn is_empty(&self) -> bool;
+
+    /// Returns the number of elements in the deque.
+    pub fn len(&self) -> usize;
+
+    /// Pushes an element onto the bottom of the deque.
+    pub fn push(&self, value: T);
+
+    /// Pops an element from the bottom of the deque.
+    pub fn pop(&self) -> Option<T>;
+
+    /// Steals an element from the top of the deque.
+    pub fn steal(&self) -> Steal<T>;
+
+    /// Creates a new stealer for the deque.
+    pub fn stealer(&self) -> Stealer<T>;
+}
+
+/// The "stealer" side of the deque.
+pub struct Stealer<T>;
+
+unsafe impl<T: Send> Send for Stealer<T> {}
+unsafe impl<T: Send> Sync for Stealer<T> {}
+
+impl<T> Stealer<T> {
+    /// Returns `true` if the deque is empty.
+    pub fn is_empty(&self) -> bool;
+
+    /// Returns the number of elements in the deque.
+    pub fn len(&self) -> usize;
+
+    /// Steals an element from the top of the deque.
+    pub fn steal(&self) -> Steal<T>;
+}
+
+impl<T> Clone for Stealer<T> { ... }
+
+/// Possible outcomes of a steal operation.
+pub enum Steal<T> {
+    /// The deque was empty.
+    Empty,
+    /// Some data was stolen.
+    Data(T),
+    /// Inconsistent state was encountered. Try again.
+    Inconsistent,
+}
+```
+
+An interesting difference from Coco's API is that a deque is now
+constructed using `Deque::new()` rather than using a global function returning
+a `(Worker<T>, Stealer<T>)` pair.
+Also, there are two ways of creating multiple stealers: you can either create each of them
+with `Deque::stealer` or by clone an existing one - whatever works best for you.
+
+Another addition is the `with_min_capacity` constructor. Deques dynamically grow and
+shrink as elements are inserted and removed. By specifying a large minimum capacity
+it is possible to reduce the frequency of reallocations.
+
+The `steal` method returns a `Steal<T>`. Instead of retrying on a failed CAS operation, the
```

Comment:
I'm slightly inclined to like the following interface better than `Steal`:

```rust
enum StealError {
    LostRace,
}

fn Stealer<T>::steal(&self) -> Result<Option<T>, StealError>;
```

Here, `Ok(Some(val))` means it returns the value `val` and `Ok(None)` means the deque is empty; `Err(LostRace)` means the thread couldn't get any useful information from the deque since it lost a race over the deque.  I think this application of `Result` and `Option` types aligns with the intended use cases of these types.

A drawback is it's a little bit verbose.

What do you think?

Reply:
Hmm, I don't know - both look fine to me. Let's see what others think.

However, one aspect of your method signature feels a bit unusual. Typically, the empty case is considered to be an error, not a special case of the success case. For example, in the `mpsc` channel, [`try_recv`](https://doc.rust-lang.org/nightly/std/sync/mpsc/struct.Receiver.html#method.try_recv) method has the following signature:

```rust
pub fn try_recv(&self) -> Result<T, TryRecvError>;

pub enum TryRecvError {
    Empty,
    Disconnected,
}
```

So the `Ok` case happens when an element is successfully received, and the `Err` case happens when the channel is either empty or disconnected. In your signature, `steal` results with `Ok` even if the deque is empty.

Reply:
@mitchmindtree Thanks for chiming in - you're presenting a very good argument!

While I agree that returning `Ok(None)` instead of `Err(Empty)` would make implementing some patterns easier, it's still slightly odd to return an `Ok` when a receive operation fails.

I mean, `try_recv` didn't *recv* a message, but it still returned an `Ok`? Hmmm... :)

Also, what if you want to receive a message from a channel, while asserting that it's not disconnected nor empty? You'd have to write:

```rust
let msg = rx.try_recv().unwrap().unwrap();
```

Aren't there too many layers of wrapping?

Perhaps this is a sign that we should keep the `Steal` enum instead of returning a `Result`? The `steal` operation is pretty unusual, so enumerating the entire list of cases on each call would make the intent of the code more obvious.

Furthermore, concurrent deques and, consequently, calls to `steal` are not very common in Rust programs (unlike channels and `try_recv`), so I believe there isn't a strong incentive to make the method as ergonomic as possible.

Reply:
> I mean, try_recv didn't recv a message, but it still returned an Ok? Hmmm... :)

Hmmm I find it interesting that you bring this up as a point against, as this seems like the intuitive behaviour to me 😄 

In my view, `try_recv` exists in order to avoid blocking when no values are present, so receiving `None` is often the expected, "successful" (`Ok`) result - I haven't personally come across a case yet where I'd describe the channel being empty as an "error" in the flow of my code or unexpected behaviour, as it seems to me this is the point of using `try_recv` instead of `recv` in the first place.

> `let msg = rx.try_recv().unwrap().unwrap();`
>
> Aren't there too many layers of wrapping?

In this case I'd argue that perhaps the user should be using `recv` if they are expecting a value to exist and `panic!`ing otherwise, e.g. `rx.recv().unwrap()`?

On the other hand, if a user really did come across a use-case for this, I'd imagine this would be more common in practise `rx.try_recv()?.unwrap()`.

> Perhaps this is a sign that we should keep the Steal enum instead of returning a Result?

Admittedly I'm not very familiar with this proposal - I just happened to come across your reference to the `try_recv` API and it sparked some frustrating memories :) Perhaps I'd be better off raising this as an alternative ergonomic approach in #22?